### ISSUE TITLE
Implement learning rate finder

### ISSUE THREAD

**Issue by matt-gardner** [allenai/allennlp#537]
This looks like a pretty good idea that would be relatively simple to implement: https://medium.com/@surmenok/estimating-optimal-learning-rate-for-a-deep-neural-network-ce32f2556ce0.

**Comment by sai-prasanna** [comment]
I can take a stab at this. Where would the lr_finder fit in the pipeline?

Could entry point be in the train command, called before training starts?

**Comment by matt-gardner** [comment]
Overall, it looks great.  Two high-level thoughts: (1) I don't much like putting the logic for the search inside of `Trainer` - it feels like it should belong in the command file that you're adding.  We should probably make some methods and fields public so that you can access them from the command.  So what we should do here is find the minimal set of things that need to be made public in order to move the logic out of `Trainer`, and do that.  (2) I'd use `find-lr` or `find-learning-rate` instead of `lr-find`.

**Comment by qiuwei** [comment]
Since allennlp is already using tensorboard, I guess plot the lr graph using tensorboard might be more convenient.

**Comment by sai-prasanna** [comment]
@matt-gardner  - I moved the logic to the find-lr.py file, But have a doubt on whether making trainer parameters public is necessary, since the find-lr is also part of allennlp package can't we access the underscore params directly? Making Trainer parameters public would imply people using the package can access those parameters without fear of change (atleast until deprecation)  and if that's okay I would make that change and give a PR.


@qiuwei  Is there API for plotting arbitary x vs y in in tensorboard?
And in future, this finder can be extended with multiple plots for different momentum or weight decay. If tensorboard can support those scenarios too, won't be much effort to change those.

**Comment by matt-gardner** [comment]
No, python doesn't really have a notion of "private", or "package protected", or whatever; we just use convention and pylint to say that some fields shouldn't be accessed outside of calls to `self._whatever`.  I'm guessing there are just a few fields that you really need access to, and those fields are probably quite stable and fine to make public, and that's preferable to putting logic inside of `Trainer` that doesn't really belong there.
